### PR TITLE
Prepare modernization

### LINKED ISSUE #1: Prepare modernization and refactoring
Prepare the modernization and refactoring by a few preparational changes. These should be conducted first so that the less relevant changes are done before the actual ones (be a friend of `git blame`).

* adjust the naming of the source files
* introduce a common coding style and enforce it by means of a `.clang-format` definition
* introduce Meson (target build system for after the modernization) and CMake (for IDE usage) build systems to the project

### PR DESCRIPTION
**Pull request by markuspg**
This prepares the further modernization and refactoring by conducting a few general changes and adjustments:

* add a `.clang-format` file for future automatic and unified formatting of the codebase
* add CMake (for IDE integration) and Meson (shall replace Autotools in future) build system files
* allow to append a version suffix to the version string
* bump the version to `2.5.0_dev` to highlight new development
* unify and adjust the naming of the headers and source files

### CODE REVIEW

File: README
Comment (linas) [inline README:30]: You moved this back to README from README.md, but URL's of this style won't work in ordinary README's.

File: configure.in
Comment (linas) [inline configure.in:2]: Your changes are so sweeping, I (strongly) recommend calling it version 3.0, not 2.5.

File: src/Makefile.am
Comment (linas) [inline src/Makefile.am:130]: You want `-lm` to be last, not first. The linker-loader resolves symbols in order.

File: redhat/gnotime.spec
Comment (linas) [inline redhat/gnotime.spec:2]: Version number here, too.

**Review by linas — COMMENTED** [review]
The explicit comments above are all I have.

Several meta-comments, which you are free to ignore, as this is now your project ... but they still leave my mind wondering....

* Why rename all the files? What does this accomplish? To me, it seems like a lot of work, nothing gained, and a complete erasure of git history making detailed reviews literally impossible.  This is the kind of thing crackers want to do, if they are tampering with code, planting back-doors or cracks, and wish to hide evidence of tampering.  So this is a big wtf...??? for any outside observer.  It makes security audits impossible.
* Same remark for the tab->space conversion.   It obfuscates changes to the code.

Meta: why spaces? If you will excuse me, but I've never met a person who actually likes to use spaces for indents. It seems very weird and un-natural to me, to have to hit the space bar a dozen or more times, when two or three whacks of the tab key are enough. I do maintain other code that has spaces in it, but I absolutely hate it.  Why would anyone ever uses spaces for indents, ever?  Is there some technical reason that I don't understand?  Is there something more to this, than a nervous thumb, or weird keyboards which move the space-bar to where the thumb is? It just seems so ergonomically awkward, un-natural.

**Review by linas — APPROVED** [review]
Change review status to "approve", but please do bump the version number to 3.0

File: configure.in
Comment (oskarb) [reply configure.in:2]: Agree, this work sounds like 3.0.

**Comment by oskarb** [comment]
I'll try to take a look at this now. I have not yet landed in any opinions, let me just make a few quick comments in reference to Linas' thoughts:

> Why rename all the files? What does this accomplish? To me, it seems like a lot of work, nothing gained, and a complete erasure of git history making detailed reviews literally impossible.

I agree that such a move should be done carefully, if at all. I have done such a complete reformatting in the past in a huge project, and in *that* case there was already a lot of mixed indenting and also mixed line-endings - it was a mess dealing with that in every PR and in the end it was better to just do a couple large pure reindenting commits to get it out of the way. If it's already consistent, there may be less reason to change.

The way @markuspg did it in this PR is that there are actually separate commits for renaming and reindenting. So looking at the diff for the renaming commits, git does produce a useful diff proving the files were renamed with few content changes. And looking at the diff for the reindent, because there are not renames in that commit, you do get a proper file by file diff (and with a diff tool that ignores indenting changes, it's manageable to check if there were any inappropriate changes "sneaked in").

If this should be done (and I have not yet landed in my own opinion yet), I definitely think that separate commits like that is the only reasonable way of doing it - which also means those commits must remain separate in main history, i.e. no "squash merge".

@markuspg, am I correct in thinking that the renames to have a file prefix has something to do with build systems using that prefix for some auto-detection? I seem to recall seeing such abilities somewhere (in general) but I haven't checked in detail.


> tabs versus spaces

Personal preference here I think may depend on what background you come from and what tools you prefer. Many IDEs, properly configured, will automatically put the cursor in the correct positions and also auto-reindent when inserting `;`, etc., so you are unlikely to ever insert indenting characters manually. The experience greatly depends on how smart the auto-indenter is.

The thing with tabs for indenting is that you also have to use spaces if you want to _align_ in multi-line statements, (even when the alignment is larger than the tab-width) otherwise it will be messed up if the code is displayed using a different tab width (which is often given as a benefit of using tabs). In the projects I've been on where tabs was used, this was often a source of confusion leading to messiness, especially for less experienced developers. Frequently PR looks unreadable because the review interface uses a different tab width, messing up alignment.

Personally I think I could live with either tabs or spaces for indenting, as long as spaces are consistently used for alignment, but I must say I find the choice of just 2 positions indent width very small. 8 is not uncommon, but my personal preference is 4, as a balance between readability and total line length.

File: README
Comment (markuspg) [reply README:30]: Generally I would prefer `README.md` too, but for some reason the new name broke the Autotools build and I was not able to figure out why, that's why I moved it back to plain `README`. If you can throw me a bone how to fix the broken build with `README.md`, I gladly go with that.

File: configure.in
Comment (markuspg) [reply configure.in:2]: I was in two minds. Increasing the dependency versions (especially GTK+ 2 to GTK+ 3) feels like a reason to bump the major version. On the other hand my intension is to keep the user facing changes (behavior and features) to zero at this stage. That's why I opted for a mere `2.5.0`. But I'm not very fixed on that, so I'll raise it to `3.0.0` then :-)

File: src/Makefile.am
Comment (markuspg) [reply src/Makefile.am:130]: Thank you for the hint, I'm not very knowledgable in this realm and will fix this accordingly.

File: configure.in
Comment (oskarb) [reply configure.in:2]: It's been 10 years! I think we deserve 3.0 :)

**Comment by markuspg** [comment]
Hi @linas, @oskarb,

thank you for your review and comments. Before you "waste" more time I want to respond quickly. I plan to redo this pull request, but in the next few days there won't be much time for this most probably.

I was wrestling with myself a bit too about changing the filenames and formatting. My feeling was that the filenames were not in all cases a good fit for the contents. Additionally I prefer "name spacing" sources too. On the other hand I myself frown upon renaming stuff versioned by Git. That's why I tried very hard not to change and rename a file in the same commit, to help Git trailing the changes ;-) Still I see your point with the obfuscation which I want to avoid too, so I guess I'll revert the change of the names.

The other "obfuscation" is the formatting. I strongly prefer using and enforcing `clang-format` to alleviate our brains from having to care about the formatting and having a unified formatting across the entire code base. Since Gnome seems to default to the "GNU" style that was my approach too. On the other hand just enforcing "GNU" style at once is really messy. I guess I'll adjust the `.clang-format` file to better fit the currently applied formatting to reduce the diff and only format future changes. By this way the formatting would get unified over time without "breaking" it all at once. Although this brings the major downside that for a long time in future commits formatting and functionality changes would be mixed. I would hope if I manage to create a `.clang-format` file fitting the current codebase well, that this problem could be reduced to a minor nuisance. Does anyone know an automated way to derive a `.clang-format` file from an existing codebase?

File: README
Comment (oskarb) [reply README:30]: Here are some suggested solutions:

https://stackoverflow.com/questions/15013672/use-autotools-with-readme-md

I guess the choice between "foreign" versus the one-line in the makefile is if we want to keep the other checks associated with the "gnu" profile or not (and I suppose we do, if the README was the only violation):

https://www.gnu.org/software/automake/manual/html_node/Strictness.html

**Comment by oskarb** [comment]
In that case, I would suggest something like this:

 * Can we have a PR that deals with the small changes like version number, and perhaps the addition of CMake and Meson?
 * If we go ahead with renaming and formatting, then that is 1 (or 2) separate PR?

For the formatting, is the existing formatting fairly consistent? If it is and if we can live with those choices, I guess we can try to stick with it. If there are many discrepancies, or if we do want to change it due to personal preferences, I would say that it's better to do all that at one time instead of having the history constantly polluted by gradual reformatting.

Btw, here is an example of what I meant with the alignment:

```c
static void
start_file_save_timer (void)
{
	g_return_if_fail (!file_save_timer);
	file_save_timer = g_timeout_add_seconds (config_autosave_period,
											 file_save_timer_func, NULL);
}
```

The third line of the function should have the same *indent* level as the second line (then aligned to the parameter) - but because tabs have been used for aligning the second parameter, by someone with a local tab width of 4, it looks misaligned in github or anywhere else where a tab width of 8 is default.


We should also have a .gitattributes file. I will look into that.

**Comment by oskarb** [comment]
@markuspg Would you upgrade my membership so I can push some branches directly to this repo?

**Comment by markuspg** [comment]
> @markuspg Would you upgrade my membership so I can push some branches directly to this repo?

Is it possible to have roles between `Member` and `Owner` on an organization or does this work with paid plans only? I had a look at it today and couldn't accomplish it :-/ Unfortunately I'm kept busy by other stuff currently, I'll have another look soon. I don't get why the pushing of branches should be blocked. The only thing I tried to block is pushing to `master`.

**Comment by linas** [comment]
As a general rule, everyone should clone the repo, push to branches in their own repo, and then create pull requests to this repo. The pull reqs sit here and can be reviewed, commented merged as needed. 
 
Periodically, each individual pulls from master on this repo to stay up-to-date

**Comment by oskarb** [comment]
@markuspg,

Right, forget what I said about being org owner for pushing to the repo - that's separate controls. I don't currently have enough access to see those controls, so I forgot about them, but I realize now I see them in repos where I've been involved earlier.

I do still think the organization roles should be tweaked - see below.


Repository access
----

I think organization owners by default has full access to all repositories, but otherwise permissions are granted to individual members (or teams) specifically for each repository. Did you set anything up there?

I think in some other projects we've used the model that core team could push directly to the main repo (but still typically do a PR for review etc), but pushing to the personal repo and PR from there might be a good idea.


Organization roles
----

Github documentation for the roles:
[Roles in an organization](https://docs.github.com/en/organizations/managing-peoples-access-to-your-organization-with-roles/roles-in-an-organization)


There is nothing between "Owner" and "Member" in terms of the basic roles. There seems to be a lot of things only Owner can do. **Crucially, it should be noted that Owners are the only ones that can add/remove members.** Partly for this reason, github does recommend that there is always a few, but multiple, owners, to reduce the risk that the organization becomes "lost" if a single owner becomes unresponsive:
[Maintaining ownership continuity for your organization](https://docs.github.com/en/organizations/managing-peoples-access-to-your-organization-with-roles/maintaining-ownership-continuity-for-your-organization)

Perhaps, since he is still responsive, and due to his long history with the project, at least @linas should be made owner of the organization. For "backup purposes" at the very least. I'll leave it up to you if you want to make me an org owner or not at this stage. :)

**Comment by markuspg** [comment]
Hi @linas, hi @oskarb,

I have paused my effort on this pull request (and the others) for now. I see your points about the renaming of the source files and the global formatting change. Right now I'm trying to code a little tool (Genetic algorithm optimizing the size of the diffs of different `.clang-format` configurations) to automatically derive `.clang-format` definitions for existing codebases and would like to use it for this project. By this I would hope we can establish a consistent and automatic source formatting which does not result in a huge diff and still helps us to free our brains from the task of formatting the code.

The renaming of the source files I would drop entirely.

**Comment by markuspg** [comment]
I'll abandon this pull request since it's outdated and most of its changes have been realized by other pull requests already.

**Pull request closed without merging**
